INT-3997: Create H2ChannelMessageStoreQueryProvider

manueljordan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3997
- Created the `H2ChannelMessageStoreQueryProvider` class for the
  `ChannelMessageStoreQueryProvider` interface
- The `LongRunningIntegrationTest` class needs the
  `RUN_LONG_INTEGRATION_TESTS` variable defined in the 'system'
  and set in true. It to let work the test methods for the
  `H2TxTimeoutMessageStoreTests` class

manueljordan committed:
INT-3997: Create H2ChannelMessageStoreQueryProvider

JIRA: https://jira.spring.io/browse/INT-3997

* Created the `H2ChannelMessageStoreQueryProvider` class for the
`ChannelMessageStoreQueryProvider` interface
* The `LongRunningIntegrationTest` class needs the
`RUN_LONG_INTEGRATION_TESTS` variable defined in the 'system'
and set in true. It to let work the test methods for the
`H2TxTimeoutMessageStoreTests` class

artembilan commented:
Hello, @manueljordan !

Thank you for contribution as usual!

I've just pushed the fix for the failure in Travis: https://github.com/spring-projects/spring-integration/commit/49985df8580e296b7aa4aed9e32032537c8e0c0b. So, please, rebase your branch to the latest master to eliminate that problem. Meanwhile I'm reviewing your PR.

artembilan commented:
H-m. I see only copy/pastes, but, yeah, no choice. To be consistent we should provide it anyway for future unpredictable cases when we need to modify one solution (H2) but don't touch another (HSQL).

My main concern here, @manueljordan .
You have raised so much JIRAs about our Reference Manual, but I don't see any particular reflection there for this your new feature :smile: 

@garyrussell , I find this fix is so simple that it can be included even into GA.

manueljordan commented:
Hello @artembilan 

> H-m. I see only copy/pastes, but, yeah, no choice

For this case, yes. That was one point I wanted to talk in the the JIRA issue, well it is the unique way *but* from here comes my second point and observation is that practically all these DB technologies (i.e Mysql, Derby, H2 and Posrgresql) reflected in their own classes have the same methods and same content, so why not create a new super class in their hierarchy?. The unique exception is the HSQL class, where it has a method to do an insertion (it if my memory does not fail me now)

Let me know if my idea is not clear.

> To be consistent we should provide it anyway for future unpredictable cases when we need to modify one solution (H2) but don't touch another (HSQL)

Please, expand the idea, I don't understand your point.

> You have raised so much JIRAs about our Reference Manual 

Well, I am doing a deep research in many components that's why I detect some doubts and other things. New features for the future, but now just doing a research in the current content available. 

> but I don't see any particular reflection there for this your new feature

If I understood well, about this new H2 support, where should be indicated or mentioned in the reference documentation? (I am assuming in many places)

According with your answers I will consider do the rebase, to see if I should touch the reference documentation or not.

Thanks

garyrussell commented:
>@garyrussell , I find this fix is so simple that it can be included even into GA.

@artembilan yes; I agree; my general rule is "new stuff": can be added after the RC, forbidden (generally - unless a bug) are schema changes and/or functional changes, I would even think schema additions are allowed.

artembilan commented:
Not sure about "the same content", because I see this in the Derby class:
````java
public String getPollFromGroupExcludeIdsQuery() {
		return "SELECT %PREFIX%CHANNEL_MESSAGE.MESSAGE_ID, %PREFIX%CHANNEL_MESSAGE.MESSAGE_BYTES from %PREFIX%CHANNEL_MESSAGE " +
				"where %PREFIX%CHANNEL_MESSAGE.GROUP_KEY = :group_key and %PREFIX%CHANNEL_MESSAGE.REGION = :region " +
				"and %PREFIX%CHANNEL_MESSAGE.MESSAGE_ID not in (:message_ids) order by CREATED_DATE, MESSAGE_SEQUENCE FETCH FIRST ROW ONLY";
}
````
Meanwhile the same method for MySQL is:
````java
public String getPollFromGroupExcludeIdsQuery() {
		return "SELECT %PREFIX%CHANNEL_MESSAGE.MESSAGE_ID, %PREFIX%CHANNEL_MESSAGE.MESSAGE_BYTES from %PREFIX%CHANNEL_MESSAGE " +
				"where %PREFIX%CHANNEL_MESSAGE.GROUP_KEY = :group_key and %PREFIX%CHANNEL_MESSAGE.REGION = :region " +
				"and %PREFIX%CHANNEL_MESSAGE.MESSAGE_ID not in (:message_ids) order by CREATED_DATE, MESSAGE_SEQUENCE LIMIT 1";
}
````
Even if there is something similar, sometimes it is really much easier to keep content in different places to have ability to modify one of them do not touching another.
That is something from the principle "composition VS inheritance".

That's why I'd prefer to leave it as is and don't spend time for the complications with an additional hierarchy level.

For this PR I'd prefer to see as an addition as a short note in the "What's New" with a link from there into the http://docs.spring.io/spring-integration/reference/html/jdbc.html#jdbc-message-store-channels, where you'll just add `H2` into the supported list.
See `whats-new.adoc` and `jdbc.adoc`
That's all.

manueljordan commented:
Hi @artembilan 

Interesting, I did a comparison between many methods and for each line... and funny you have found a difference... well you are the expert here...

BTW: I have taken how reference the content from HSQL to create the H2.

Understood about the reference documentation, let me do it tomorrow early morning with my mind totally relaxed ...

Best Regards!

manueljordan commented:
Practically completed, but to avoid an unnecessary commit, for the `whats-new.doc` file.

Is ok for you if for the `General Changes` section I add the following _how latest item_?:

**Jdbc Changes**

**Backing Message Channels**

**General**

The `ChannelMessageStoreQueryProvider` now supports `H2` database. See "Backing Message Channels" for more details

Let me know if is ok for you.

Thanks

artembilan commented:
I think there would be enough just call a new short section as `ChannelMessageStoreQueryProvider`.
See `@InboundChannelAdapter` section for example.

manueljordan committed:
INT-3997: Create H2ChannelMessageStoreQueryProvider

JIRA: https://jira.spring.io/browse/INT-3997

* Updated reference documentation about
`ChannelMessageStoreQueryProvider`

artembilan commented:
@manueljordan , it would be good to let us know that you have pushed a new commit.
GH doesn't notify about changes.
Just simple `Pushed` comment is enough. Everything else we can pick up from the commit comments (can we ? BTW :wink: ) and code changes.

Thank you for the contribution anyway!

artembilan commented on spring-integration-jdbc/src/main/resources/org/springframework/integration/jdbc/store/channel/schema-h2.sql:
```diff
@@ -0,0 +1,15 @@
+CREATE INDEX INT_CHANNEL_MSG_DATE_IDX ON INT_CHANNEL_MESSAGE (CREATED_DATE, MESSAGE_SEQUENCE);
+CREATE INDEX INT_CHANNEL_MSG_PRIORITY_IDX ON INT_CHANNEL_MESSAGE (MESSAGE_PRIORITY DESC, CREATED_DATE, MESSAGE_SEQUENCE);
+
+CREATE SEQUENCE INT_MESSAGE_SEQ START WITH 1 INCREMENT BY 1;
```
Your IDE is still have vulnerabilities with "New line in the end" :smile: 
Will fix on merge.

manueljordan replied:
I have open that file, there is no an extra blank line after of the latest line, weird... my IDE shows just 15 lines too..

artembilan commented on src/reference/asciidoc/whats-new.adoc:
```diff
@@ -260,3 +260,7 @@ See <<xmpp-extensions>> for more information.
+
+==== ChannelMessageStoreQueryProvider
+
+The `ChannelMessageStoreQueryProvider` now supports H2 database. See <<jdbc-message-store-channels>> for more information.
```
Minor too, but need to know as well: AsciiDoc [recommendations](http://asciidoctor.org/docs/asciidoc-recommended-practices/) to start a new Sentence from a new line.

H-m, @garyrussell , we still wrap our sentences on the 120 length. Should we break this rule for our docs? :smile:

manueljordan replied:
I am sorry, I can see I have more of 120 columns. but, what do you mean with "start a new Sentence from a new line."?

garyrussell replied:
`This is one sentence.`
`This is another.`
`This is two sentences. Which are on the same line`.

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/c6821ec3875bde148131e29a0e07fe5a166a3a24

garyrussell commented:
>Don’t wrap text at a fixed column width. Instead, put each sentence on its own line, a technique called sentence per line. This technique is similar to how you write and organize source code. The result can be spectacular.

Interesting yes; I am fine with keeping longer lines (one sentence per line) - github diffs seem to handle it ok (unlike Java code).

The only downside I can think of is `git diff` on the command line gives truncated lines - but it seems only I use command-line git 😄
